nomad/util.go

```diff
@@ -148,7 +148,29 @@ func isNomadServer(m serf.Member) (bool, *serverParts) {
 // for the failed servers should be verified.
 func ServersMeetMinimumVersion(members []serf.Member, minVersion *version.Version, checkFailedServers bool) bool {
 	for _, member := range members {
-		if valid, parts := isNomadServer(member); valid && (parts.Status == serf.StatusAlive || (checkFailedServers && parts.Status == serf.StatusFailed)) {
+		valid, parts := isNomadServer(member)
+		if valid && (parts.Status == serf.StatusAlive || (checkFailedServers && parts.Status == serf.StatusFailed)) {
+			// Check if the versions match - version.LessThan will return true for
+			// 0.8.0-rc1 < 0.8.0, so we want to ignore the metadata
+			versionsMatch := slicesMatch(minVersion.Segments(), parts.Build.Segments())
+			if parts.Build.LessThan(minVersion) && !versionsMatch {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+// RegionServersMeetMinimumVersion returns whether the Nomad servers in this
+// region are at least on the given Nomad version. The checkFailedServers
+// parameter specifies whether version for the failed servers should be
+// verified.
+func RegionServersMeetMinimumVersion(members []serf.Member, region string, minVersion *version.Version, checkFailedServers bool) bool {
```

Comment:
Since this is the same as `ServersMeetMinimumVersion` I'm tempted to suggest just expanding that functions parameter list to include `region string`, which matches all regions if unset.

Reply:
While we're in here we can probably kill `slicesMatch` for `helper.SliceSetEq`

Reply:
> Since this is the same as `ServersMeetMinimumVersion` I'm tempted to suggest just expanding that functions parameter list to include `region string`, which matches all regions if unset. 

We have a whole bunch of call sites for this for legacy bits I didn't want to tear out while trying to fix a bug at the same time. If we go and touch those now, we'd want to change the behavior too, otherwise it's going to look like we intentionally set them to check all regions. I was intentionally not touching those but honestly it's probably safe to go ahead and do it at this point. Will do.

> While we're in here we can probably kill `slicesMatch` for `helper.SliceSetEq`

Isn't `SliceSetEq` specifically for a _set_ (i.e. unordered). The segments of a version have meaningful order.

Reply:
Turns out I can't mess with this signature without breaking all the backports in the process. So I'm going to rework all the open PRs for this problem into one PR per backport.

Reply:
oh derp, yeah. in that case there is https://pkg.go.dev/golang.org/x/exp/slices#Equal but yeah that's just optional cleanup

Reply:
You could also just keep both public signatures and have them call into a shared private helper

Reply:
I've squooshed it all into https://github.com/hashicorp/nomad/pull/14912 and I'll rework this PR to just cover 1.2.x and https://github.com/hashicorp/nomad/pull/14911 to just cover 1.3.x.